### PR TITLE
fix: Exceeding quota with volumeClaimTemplates

### LINKED ISSUE #3419: Exceeding quota with volumeClaimTemplates
Checklist:

 * [x] I've included the version.
 * [x] I've included reproduction steps.
 * [x] I've included the workflow YAML.
 * [x] I've included the logs.
 
**What happened**:
I have created a template which uses a volume claim template. If the total claimed by running pods is at storage quota then new pods being attempted to run enter error state rather than pending state.

**What you expected to happen**:
If storage resources are unavailable then workflow should remain pending until resources are available.

**How to reproduce it (as minimally and precisely as possible)**:
In a namespace with a quota of 10Gi of storage the error should start occuring after 5 have been submitted.
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: storage-limit-
spec:
  serviceAccountName: argo
  entrypoint: wait
  podGC:
    strategy: OnPodCompletion

  volumeClaimTemplates:                 # define volume, same syntax as k8s Pod spec
    - metadata:
        name: workdir                     # name of volume claim
      spec:
        accessModes: [ "ReadWriteMany" ]
        resources:
          requests:
            storage: 2Gi                  # Gi => 1024 * 1024 * 1024

  templates:
  - name: wait
    resubmitPendingPods: true
    script:
      image: alpine:latest
      command: [sh, -c]
      args: ["sleep", "30s"]
```
Use bash script to run this some number of times until you encounter the issue.

**Anything else we need to know?**:
I believe this issue also occurs when hitting PVC limits.

**Environment**:

- Argo version:
```
$ argo version
argo: 2.8.2+8a151ae.dirty
  BuildDate: 2020-06-18T23:50:58Z
  GitCommit: 8a151aec6538c9442cf2380c2544ba3efb60ff60
  GitTreeState: dirty
  GitTag: 2.8.2
  GoVersion: go1.13
  Compiler: gc
  Platform: linux/amd64
```
- Kubernetes version :
```
$ kubectl version -o yaml
clientVersion:
  buildDate: 2020-01-29T21:26:39Z
  compiler: gc
  gitCommit: d4cacc0
  gitTreeState: clean
  gitVersion: v1.10.0+d4cacc0
  goVersion: go1.14beta1
  major: "1"
  minor: 10+
  platform: linux/amd64
serverVersion:
  buildDate: 2020-05-04T12:54:43Z
  compiler: gc
  gitCommit: a3ec9df
  gitTreeState: clean
  gitVersion: v1.16.2
  goVersion: go1.12.12
  major: "1"
  minor: 16+
  platform: linux/amd64
```

**Other debugging information (if applicable)**:

- workflow result:

```
$ argo --loglevel DEBUG get <workflowname>
DEBU[0000] CLI version                                   version="{2.8.2+8a151ae.dirty 2020-06-18T23:50:58Z 8a151aec6538c9442cf2380c2544ba3efb60ff60 2.8.2 dirty go1.13 gc linux/amd64}"
DEBU[0000] Client options                                opts="{{ false false} 0x1574670 0xc00011f040}"
Name:                storage-limit-v9qjt
Namespace:           thoth-test-core
ServiceAccount:      argo
Status:              Error
Message:             persistentvolumeclaims "storage-limit-v9qjt-workdir" is forbidden: exceeded quota: thoth-test-core-quota, requested: requests.storage=2Gi, used: requests.storage=40Gi, limited: requests.storage=40Gi
Conditions:          
 Completed           True
Created:             Wed Jul 08 11:57:57 -0400 (7 minutes ago)
Started:             Wed Jul 08 11:57:57 -0400 (7 minutes ago)
Finished:            Wed Jul 08 11:57:57 -0400 (7 minutes ago)
Duration:            0 seconds
```

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by sarabala1979**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  Fixes #3419 
* [ ] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -1243,6 +1252,9 @@ func (woc *wfOperationCtx) createPVCs() error {
 		if pvcTmpl.ObjectMeta.Name == "" {
 			return errors.Errorf(errors.CodeBadRequest, "volumeClaimTemplates[%d].metadata.name is required", i)
 		}
+		if found := existPVC[pvcTmpl.ObjectMeta.Name]; found {
```
Comment (alexec) [inline workflow/controller/operator.go:1255]: delete these changes as you do not need them

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -1225,15 +1233,16 @@ func inferFailedReason(pod *apiv1.Pod) (wfv1.NodePhase, string) {
...
+	if !(woc.wf.Status.Phase == wfv1.NodePending || woc.wf.Status.Phase == wfv1.NodeRunning) {
+		// Only attempt to create PVCs if workflow is in Pending or Running state
 		// (e.g. passed validation, or didn't already complete)
 		return nil
 	}
-	if len(woc.wfSpec.VolumeClaimTemplates) == len(woc.wf.Status.PersistentVolumeClaims) {
-		// If we have already created the PVCs, then there is nothing to do.
-		// This will also handle the case where workflow has no volumeClaimTemplates.
-		return nil
+	existPVC := make(map[string]bool)
```
Comment (alexec) [inline workflow/controller/operator.go:1241]: delete these changes as you do not need them

File: workflow/controller/operator.go [hunk 3]
```diff
@@ -282,6 +282,14 @@ func (woc *wfOperationCtx) operate() {
 
 	err = woc.createPVCs()
 	if err != nil {
+		if apierr.IsForbidden(err) {
+			// Error was most likely caused by a lack of resources.
+			// In this case, Workflow will be in pending state and requeue.
+			woc.markWorkflowPhase(wfv1.NodePending, false)
+			woc.requeue(0)
```
Comment (alexec) [inline workflow/controller/operator.go:289]: `woc.requeue(10 * time.Second)`

File: test/e2e/functional_test.go [hunk 4]
```diff
@@ -639,6 +639,24 @@ spec:
...
+	s.Given().
+		Workflow("@testdata/storage-limit.yaml").
+		When().
+		StorageQuota("5Mi").
+		SubmitWorkflow().
+		WaitForWorkflowToStart(5*time.Second).
+		WaitForWorkflowCondition(func(wf *wfv1.Workflow) bool {
+			return wfv1.NodePending == wf.Status.Phase
+		}, "PVC pending", 10*time.Second).
+		DeleteQuota().
```
Comment (alexec) [inline test/e2e/functional_test.go:652]: `DeleteStorageQuota`

File: test/e2e/functional_test.go [hunk 5]
```diff
@@ -639,6 +639,24 @@ spec:
...
 
+func (s *FunctionalSuite) TestStorageQuotaLimit() {
+	s.Given().
+		Workflow("@testdata/storage-limit.yaml").
+		When().
+		StorageQuota("5Mi").
+		SubmitWorkflow().
+		WaitForWorkflowToStart(5*time.Second).
+		WaitForWorkflowCondition(func(wf *wfv1.Workflow) bool {
+			return wfv1.NodePending == wf.Status.Phase
```
Comment (alexec) [inline test/e2e/functional_test.go:650]: assert on message

File: test/e2e/functional_test.go [hunk 6]
```diff
@@ -639,6 +639,24 @@ spec:
...
+		}, "PVC pending", 10*time.Second).
+		DeleteQuota().
+		WaitForWorkflow(30 * time.Second).
+		Then().
+		ExpectWorkflow(func(t *testing.T, _ *metav1.ObjectMeta, status *wfv1.WorkflowStatus) {
+			assert.Equal(t, wfv1.NodeSucceeded, status.Phase)
```
Comment (alexec) [inline test/e2e/functional_test.go:656]: assert message gone

File: workflow/controller/operator.go [hunk 7]
```diff
@@ -282,6 +282,14 @@ func (woc *wfOperationCtx) operate() {
 
 	err = woc.createPVCs()
 	if err != nil {
+		if apierr.IsForbidden(err) {
+			// Error was most likely caused by a lack of resources.
+			// In this case, Workflow will be in pending state and requeue.
+			woc.markWorkflowPhase(wfv1.NodePending, false)
```
Comment (alexec) [inline workflow/controller/operator.go:288]: please add the reason:

```
"Failed to create PVCs (maybe due to lack of resources): %s", err)
```

File: workflow/controller/operator.go [hunk 8]
```diff
@@ -1290,7 +1295,11 @@ func (woc *wfOperationCtx) createPVCs() error {
 				},
 			},
 		}
-		woc.wf.Status.PersistentVolumeClaims[i] = vol
+		woc.wf.Status.PersistentVolumeClaims = append(woc.wf.Status.PersistentVolumeClaims, vol)
+		// Clearing previous PVC forbidden error
+		if woc.wf.Status.Message != "" && strings.Contains(woc.wf.Status.Message, "Waiting for a PVC to be created") {
```
Comment (alexec) [inline workflow/controller/operator.go:1300]: minor - extract "Waiting for a PVC to be created" to a const and use on line 288

File: workflow/controller/operator.go [hunk 9]
```diff
@@ -1673,7 +1680,7 @@ func (woc *wfOperationCtx) hasDaemonNodes() bool {
 }
 
 func (woc *wfOperationCtx) markWorkflowRunning() {
-	woc.markWorkflowPhase(wfv1.NodeRunning, false)
+	woc.markWorkflowPhase(wfv1.NodeRunning, false, "")
```
Comment (sarabala1979) [inline workflow/controller/operator.go:1683]: Previous error/pending message needs to clear on running phase

File: workflow/controller/operator_test.go [hunk 10]
```diff
@@ -4101,7 +4101,7 @@ status:
 		defer cancel()
 		woc := newWorkflowOperationCtx(wf, controller)
 		woc.operate()
-		assert.Equal(t, wfv1.NodePending, woc.wf.Status.Phase)
+		assert.Equal(t, wfv1.NodeRunning, woc.wf.Status.Phase)
```
Comment (sarabala1979) [inline workflow/controller/operator_test.go:4104]: Resubmit will memoization wf phase should be running after `operate` function. I will create separate PR to fix the phase in memorization

Comment (alexec) [reply workflow/controller/operator_test.go:4104]: ok

Comment (alexec) [reply workflow/controller/operator_test.go:4104]: if there is another bug fix - can we combine into a single valid PR?

Comment (sarabala1979) [reply workflow/controller/operator_test.go:4104]: sure I will fix the memoization bug also in this PR.

Comment (sarabala1979) [reply workflow/controller/operator_test.go:4104]: As per @jessesuen discussed, just need to fix the phase in resubmit memorized scenario.

**Pull request merged**